src/derivers/shared.ts

```diff
@@ -0,0 +1,540 @@
+  entropy,
+  publicKey,
+  depth,
+  masterFingerprint,
+  parentFingerprint,
+  childIndex,
+  curve,
+}: DerivePublicChildKeyArgs): Promise<SLIP10Node> {
+  const { publicKey: childPublicKey, chainCode: childChainCode } =
+    generatePublicKey({
+      publicKey,
+      entropy,
+      curve,
+    });
+
+  return await SLIP10Node.fromExtendedKey({
+    publicKey: childPublicKey,
+    chainCode: childChainCode,
+    depth: depth + 1,
+    masterFingerprint,
+    parentFingerprint,
+    index: childIndex,
+    curve: curve.name,
+  });
+}
+
+/**
+ * Add a tweak to the private key: `(privateKey + tweak) % n`.
+ *
+ * @param privateKeyBytes - The private key as 32 byte Uint8Array.
+ * @param tweakBytes - The tweak as 32 byte Uint8Array.
+ * @param curve - The curve to use.
+ * @throws If the private key or tweak is invalid.
+ * @returns The private key with the tweak added to it.
+ */
+export function privateAdd(
+  privateKeyBytes: Uint8Array,
+  tweakBytes: Uint8Array,
+  curve: Curve,
+): Uint8Array {
+  assert(
+    isValidBytesKey(tweakBytes, 32),
+    'Invalid tweak: Tweak must be a non-zero 32-byte Uint8Array.',
+  );
+
+  const privateKey = bytesToBigInt(privateKeyBytes);
+  const tweak = bytesToBigInt(tweakBytes);
+
+  if (tweak >= curve.curve.n) {
+    throw new Error('Invalid tweak: Tweak is larger than the curve order.');
+  }
+
+  const added = mod(privateKey + tweak, curve.curve.n);
+  const bytes = hexToBytes(added.toString(16).padStart(64, '0'));
+
+  if (!curve.isValidPrivateKey(bytes)) {
+    throw new Error(
+      'Invalid private key or tweak: The resulting private key is invalid.',
+    );
+  }
+
+  return bytes;
+}
+
+type GenerateEntropyArgs = {
+  chainCode: Uint8Array;
+  extension: Uint8Array;
+};
+
+/**
+ * Generate 64 bytes of (deterministic) entropy from a chain code and secret
+ * extension.
+ *
+ * @param args - The arguments for generating entropy.
+ * @param args.chainCode - The parent chain code bytes.
+ * @param args.extension - The extension bytes.
+ * @returns The generated entropy bytes.
+ */
+export function generateEntropy({ chainCode, extension }: GenerateEntropyArgs) {
+  return hmac(sha512, chainCode, extension);
+}
+
+/**
+ * Validate that a node is specified.
+ *
+ * @param node - The node to validate.
+ * @throws If the node is not specified.
+ */
+export function validateNode(node?: SLIP10Node): asserts node is SLIP10Node {
+  assert(node, 'Invalid parameters: Must specify a node to derive from.');
+}
+
+/**
+ * Validate a path.
+ *
+ * @param path - The path to validate.
+ * @param node - The node to validate the path against.
+ * @param curve - The curve to validate the path against.
+ * @throws If the path is invalid.
+ */
+function validatePath(
+  path: string | Uint8Array,
+  node: SLIP10Node,
+  curve: Curve,
+): asserts path is string {
+  assert(typeof path === 'string', 'Invalid path: Must be a string.');
+
+  const isHardened = path.endsWith(`'`);
```

Comment:
Shouldn't this be `includes()`?

Reply:
Hardened paths always end with `'`, which is enforced by our regex.